cxx-squid/src/main/java/org/sonar/cxx/CxxMetricsFactory.java

```diff
@@ -0,0 +1,155 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.cxx;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import org.sonar.api.measures.Metric;
+
+/**
+ * Some metrics are language specific. The exact reason for that might be the
+ * coexistence of CXX and C plugins in the same project/module. This concept
+ * doesn't match with CoreMetrics e.g. complexity, issues number etc are
+ * language independent; for multi-language modules CoreMetrics are accumulated.
+ *
+ * Unification of metrics should be discussed in the future.
+ *
+ */
+public class CxxMetricsFactory {
+
+  private CxxMetricsFactory() {
+  }
+
+  public enum Key {
+    // Introduce own documentation metrics, after they has been removed from SQ core
+    // see https://jira.sonarsource.com/browse/SONAR-8328
+    PUBLIC_API_KEY("public_api"),
+    PUBLIC_UNDOCUMENTED_API_KEY("public_undocumented_api"),
+    PUBLIC_DOCUMENTED_API_DENSITY_KEY("public_documented_api_density"),
+
+    // Introduce metric keys for sensors - number of issues per file / per module
+    CLANG_SA_SENSOR_ISSUES_KEY("ClangSA"),
+    CLANG_TIDY_SENSOR_ISSUES_KEY("Clang-Tidy"),
+    COMPILER_SENSOR_ISSUES_KEY("Compiler"),
+    CPPCHECK_SENSOR_ISSUES_KEY("CppCheck"),
+    DRMEMORY_SENSOR_ISSUES_KEY("DrMemory"),
+    OTHER_SENSOR_ISSUES_KEY("other"),
+    PCLINT_SENSOR_ISSUES_KEY("PC-Lint"),
+    RATS_SENSOR_ISSUES_KEY("Rats"),
+    SQUID_SENSOR_ISSUES_KEY("Squid"),
+    VALGRIND_SENSOR_KEY("Valgrind"),
+    VERAXX_SENSOR_KEY("Vera++");
+
+    String getLanguageSpecificKey(String langPropertiesKey)
+    {
+      return langPropertiesKey.toUpperCase(Locale.ENGLISH) + "-" + key;
+    }
+
+    Key(final String key) {
+      this.key = key.toUpperCase(Locale.ENGLISH);
+    }
+
+    @Override
+    public String toString() {
+      return key;
+    }
+
+    private final String key;
+  }
+
+  /**
+   * Generate a map
+   * <code>{ language independent metric key : language dependent Metric object }</code>
+   */
+  public static Map<CxxMetricsFactory.Key, Metric<?>> generateMap(String langKey, String langPropertiesKey) {
+    final String domain = langKey.toUpperCase(Locale.ENGLISH);
+
+    final Map<CxxMetricsFactory.Key, Metric<?>> metrics = new HashMap<>();
+
+    final Metric<Integer> PUBLIC_API = new Metric.Builder(Key.PUBLIC_API_KEY.getLanguageSpecificKey(langPropertiesKey),
+        "Public API", Metric.ValueType.INT)
+        .setDescription("Public API")
+        .setDirection(Metric.DIRECTION_WORST)
+         .setQualitative(false)

```

Comment:
_needlessly boxes a boolean constant_ (fb-contrib:NAB_NEEDLESS_BOOLEAN_CONSTANT_CONVERSION)
3 times in this file